src/Compilers/CSharp/Test/Emit3/Semantics/UnsafeEvolutionTests.cs

```diff
@@ -2,13 +2,423 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Linq;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
 using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Utilities;
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Semantics;
 
 public sealed class UnsafeEvolutionTests : CompilingTestBase
 {
+    private static void VerifyMemorySafetyRulesAttribute(ModuleSymbol module, bool includesAttributeDefinition, bool includesAttributeUse, bool publicDefinition)
+    {
+        const string name = "MemorySafetyRulesAttribute";
+        const string fullName = $"System.Runtime.CompilerServices.{name}";
+        var type = (NamedTypeSymbol)module.GlobalNamespace.GetMember(fullName);
+        var attribute = module.GetAttributes().SingleOrDefault(a => a.AttributeClass?.Name == name);
+
+        if (includesAttributeDefinition)
+        {
+            Assert.NotNull(type);
+        }
+        else
+        {
+            Assert.Null(type);
+            if (includesAttributeUse)
+            {
+                Assert.NotNull(attribute);
+                type = attribute.AttributeClass;
+            }
+        }
+
+        if (type is { })
+        {
+            Assert.Equal(publicDefinition ? Accessibility.Public : Accessibility.Internal, type.DeclaredAccessibility);
```

Comment:
Please verify that the attribute has been marked as `Embedded` when appropriate.